src/InternalEvents.tsx

```diff
@@ -196,15 +195,23 @@ function useShortcuts() {
       const shortcut = action.shortcut!.join(" ");
 
       shortcutsMap[shortcut] = wrap((event: KeyboardEvent) => {
-        if (shouldRejectKeystrokes()) return;
+        if (shouldRejectKeystrokes(shortcut)) return;
 
         event.preventDefault();
-        if (action.children?.length) {
-          query.setCurrentRootAction(action.id);
+
+        if (currentRootActionId === action.id && open) {
+          // Close kbar when user choosed current action double time.
           query.toggle();
+          options.callbacks?.onClose?.();
+        } else if (action.children?.length) {
+          query.setCurrentRootAction(action.id);
+          // When action has children toggle only when kbar is hidden.
+          !open && query.toggle();
           options.callbacks?.onOpen?.();
         } else {
           action.command?.perform();
+          // When action doesn't have children toggle only when kbar is showing.
+          open && query.toggle();
           options.callbacks?.onSelectAction?.(action);
         }
```

Comment:
hey @notmedia, this logic is getting quite complex and be it that it took me quite a while to figure out what is actually happening, can you add an example of how this works in the example / add a description and screencap of what the expected behavior is?

Reply:
Exactly, maybe we could rewrite and get it more simpler

First case - closing kbar when the user use the same shortcut twice. This worked before only with `mod+K`

https://user-images.githubusercontent.com/12993018/200199479-902b50c4-3a67-4f93-82e9-2048ff13ba6a.mov

Second case - when we have action with children  and the user use shortcut - kbar opens on this children menu

https://user-images.githubusercontent.com/12993018/200199586-3ef5c409-f7f6-4f21-bf0f-9c09cedb46eb.mov

Third case - when action does not have children we do not need to open kbar when it closed, but when it already opened we need to close it

https://user-images.githubusercontent.com/12993018/200199724-6b59823d-a5ac-4443-8b8a-ea512b53cfad.mov

Reply:
@timc1 I hope this will help)

Reply:
> First case - closing kbar when the user use the same shortcut twice. This worked before only with mod+K

I don't think this should be the expected behavior. Even with the condition that the shortcut begins with `$mod`, this creates implicit behavior that is difficult to understand (ie; why would hitting `$mod+c` handle both opening/closing kbar, and `g+d` not?)

> Second case - when we have action with children and the user use shortcut - kbar opens on this children menu

In this case, to close kbar you just hit `ESC`?

> Third case - when action does not have children we do not need to open kbar when it closed, but when it already opened we need to close it

If an action doesn't have any children, kbar shouldn't be opening in the first place – that would be a bug. When you define the action for `$mod+x` in the above video, `children` for that action should be `undefined`.

Reply:
> I don't think this should be the expected behavior. Even with the condition that the shortcut begins with $mod, this creates implicit behavior that is difficult to understand (ie; why would hitting $mod+c handle both opening/closing kbar, and g+d not?)

My thoughts was that shortcuts started with mod key are like system commands. So its okay, because its the same behavior to current mod+k

> In this case, to close kbar you just hit ESC?

No, I used mod+e on open and close. The main thing in this case that this action is not a root action, and I can open it

> If an action doesn't have any children, kbar shouldn't be opening in the first place – that would be a bug. When you define the action for $mod+x in the above video, children for that action should be undefined.

On this video I opened kbar with default mod+k and clicked to option “close tab”
On second iteration I used shortcut mod+w with opened kbar - it also closed kbar and perform actiob
Third iteration - without opened kbar hit shortcut and perform action

Reply:
> I used mod+e on open and close. The main thing in this case that this action is not a root action, and I can open it

Can you use `toggleShortcut` for your mod+e`? https://github.com/timc1/kbar/blob/70ce0c472e49ea8ead3886a92dffd329248fa882/src/types.ts#L65-L69

Reply:
@timc1 No, because this option overrides only mod+k.

I have some root actions with shortcuts with children. When kbar closed I want top open this children view.

Reply:
I may be completely misunderstanding here, but how does this differ from the `?` shortcut for docs in our example site?

Reply:
@timc1  Look, the difference here is that I added support `mod+` shortcuts in kbar and add some tunes to behaviour because before you cant use shortcut when kbar is open, also if you have action with children and you use shortcut twice - kbar will close

The main problem of kbar - it does not work when input is focused. The next things in videos I sent you before

Reply:
For example I have command `$mod+W` - standard system command for closing tabs. When I hit this tab on kbar site it closes tab, and I can't overwrite this behaviour

Reply:
Would a better solution be passing the DOM event through to [`perform`](https://github.com/timc1/kbar/blob/70ce0c472e49ea8ead3886a92dffd329248fa882/src/action/Command.ts#L21-L22
) such that:

```
const action = {
   perform: (event) => {
     event.preventDefault();
     // close in-app tab
   }
}
```

Reply:
This does not work now, we do not have event argument in this function

Reply:
This does not work now, we do not have event argument in this function

Reply:
But the idea is geat. Also we need to have option to use mod+key shortcuts in focused elements

Reply:
Right, we're trying to figure out a proper API here to solve the problem you have – if we're able to use simple web APIs to solve this problem, I'd rather us implement the ability to pass the native event through to `perform`.

Reply:
So you will do it or you want me to do this?

Reply:
@timc1 Really need this functionality :)

Reply:
Will have something up later tonight.

Reply:
@notmedia so doing a little digging around and looks like browsers aren't going to allow us to preventDefault for `cmd+w`/`ctrl+w`; you will need to create your own implementation with `beforeUnload` in order to handle that case. Here's an experiment to pipe default events through to `perform`, but I realized even calling `event.preventDefault` won't work in your case.

Reply:
Okay, could we enable mod shortcuts when input is focused?

Reply:
With this PR cmd+w works in electron by the way

Reply:
Explicitly only enabling shortcuts that begin with `$mod` to work when kbar is open is unfortunately feature creep that I don't feel confident in adding to the core library.

A better option for you would be piping `event` through to `perform` callback: https://github.com/timc1/kbar/pull/253, which _should_ work in your electron case as well. 

However, I'd suggest rethinking your `cmd+w` action and removing the shortcut as the behavior in your app won't be the same between web/desktop.